src/lib/pubkey/ec_group/ec_inner_data.cpp

```diff
@@ -75,63 +81,122 @@ void EC_Group_Data::set_oid(const OID& oid) {
-   return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), bn);
+   if(bn <= 0 || bn >= m_order) {
+      return {};
+   }
+
+   if(m_pcurve) {
+      return this->scalar_deserialize(bn.serialize(m_order_bytes));
+   } else {
+      return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), bn);
+   }
 }
 
 std::unique_ptr<EC_Scalar_Data> EC_Group_Data::gk_x_mod_order(const EC_Scalar_Data& scalar,
                                                               RandomNumberGenerator& rng,
                                                               std::vector<BigInt>& ws) const {
-   const auto& bn = EC_Scalar_Data_BN::checked_ref(scalar);
-   const auto pt = m_base_mult.mul(bn.value(), rng, m_order, ws);
-
-   if(pt.is_zero()) {
-      return scalar_zero();
+   if(m_pcurve) {
+      const auto& k = EC_Scalar_Data_PC::checked_ref(scalar);
+      auto gk_x_mod_order = m_pcurve->base_point_mul_x_mod_order(k.value(), rng);
+      return std::make_unique<EC_Scalar_Data_PC>(shared_from_this(), gk_x_mod_order);
    } else {
-      return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), mod_order(pt.get_affine_x()));
+      const auto& k = EC_Scalar_Data_BN::checked_ref(scalar);
+      BOTAN_STATE_CHECK(m_base_mult != nullptr);
+      const auto pt = m_base_mult->mul(k.value(), rng, m_order, ws);
+
+      if(pt.is_zero()) {
+         return scalar_zero();
+      } else {
+         return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), mod_order(pt.get_affine_x()));
+      }
    }
 }
 
-std::unique_ptr<EC_Scalar_Data> EC_Group_Data::scalar_deserialize(std::span<const uint8_t> bytes) {
+std::unique_ptr<EC_Scalar_Data> EC_Group_Data::scalar_deserialize(std::span<const uint8_t> bytes) const {
    if(bytes.size() != m_order_bytes) {
       return nullptr;
    }
 
-   BigInt r(bytes.data(), bytes.size());
+   if(m_pcurve) {
+      if(auto s = m_pcurve->deserialize_scalar(bytes)) {
+         return std::make_unique<EC_Scalar_Data_PC>(shared_from_this(), *s);
+      } else {
+         return nullptr;
+      }
+   } else {
+      BigInt r(bytes);
 
-   if(r.is_zero() || r >= m_order) {
-      return nullptr;
-   }
+      if(r.is_zero() || r >= m_order) {
+         return nullptr;
+      }
 
-   return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), std::move(r));
+      return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), std::move(r));
+   }
 }
 
 std::unique_ptr<EC_AffinePoint_Data> EC_Group_Data::point_deserialize(std::span<const uint8_t> bytes) const {
    try {
-      auto pt = Botan::OS2ECP(bytes.data(), bytes.size(), curve());
-      return std::make_unique<EC_AffinePoint_Data_BN>(shared_from_this(), std::move(pt));
+      if(m_pcurve) {
+         auto pt = m_pcurve->deserialize_point(bytes);
+         if(pt) {
+            return std::make_unique<EC_AffinePoint_Data_PC>(shared_from_this(), std::move(*pt));
+         } else {
+            return nullptr;
+         }
```

Comment:
For consistency with other methods:

```suggestion
         if(auto pt = m_pcurve->deserialize_point(bytes)) {
            return std::make_unique<EC_AffinePoint_Data_PC>(shared_from_this(), std::move(*pt));
         } else {
            return nullptr;
         }
```